### PR TITLE
LUCENE-8033: FieldInfos always use dense encoding

### PR DESCRIPTION
**Pull request by mayya-sharipova**
FieldInfos always to use an array to store FieldInfo byNumber

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/index/FieldInfos.java [hunk 1]
```diff
@@ -63,21 +60,30 @@ public FieldInfos(FieldInfo[] infos) {
...
     for (FieldInfo info : infos) {
       if (info.number < 0) {
         throw new IllegalArgumentException("illegal field number: " + info.number + " for field " + info.name);
       }
-      FieldInfo previous = byNumber.put(info.number, info);
+      size = info.number >= size ? info.number+1 : size;
+      if (info.number >= capacity){ //grow array
+        capacity = info.number + 1;
+        byNumberTemp = Arrays.copyOf(byNumberTemp, capacity);
+      }
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: can you grow the array exponentially so that it doesn't run in quadratic time in the worst case? ArrayUtil has some methods that make it easy: 
```
if (info.number >= byNumberTemp.length) {
  byNumberTemp = ArrayUtil.grow(byNumberTemp, info.number + 1);
}
```

File: lucene/core/src/java/org/apache/lucene/index/FieldInfos.java [hunk 2]
```diff
@@ -96,25 +102,19 @@ public FieldInfos(FieldInfo[] infos) {
...
+        if (byNumberTemp[i] != null)
+          valuesTemp.add(byNumberTemp[i]);
       }
     } else {
-      byNumberMap = byNumber;
-      values = Collections.unmodifiableCollection(byNumber.values());
-      byNumberTable = null;
+      byNumber = null;
     }
+    values = Collections.unmodifiableCollection(valuesTemp);
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: Let's maybe allow empty slots at the end of the array list to be collected by doing something like this?

```java
values = Collections.unmodifiableCollection(Arrays.asList(valuesTemp.toArray(new FieldInfo[0])));
```

File: lucene/core/src/java/org/apache/lucene/index/FieldInfos.java [hunk 3]
```diff
@@ -96,25 +102,19 @@ public FieldInfos(FieldInfo[] infos) {
...
+      for(int i=0; i<size; i++){
+        byNumber[i] = byNumberTemp[i];
+        if (byNumberTemp[i] != null)
+          valuesTemp.add(byNumberTemp[i]);
       }
     } else {
-      byNumberMap = byNumber;
-      values = Collections.unmodifiableCollection(byNumber.values());
-      byNumberTable = null;
+      byNumber = null;
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: It seems a bit trappy to use `null` to represent empty arrays, especially since the `fieldInfo` method doesn't test for nullity?

File: lucene/core/src/test/org/apache/lucene/index/TestFieldInfos.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.StringField;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.Test;
+
+public class TestFieldInfos extends LuceneTestCase {
+
+  @Test
```
Comment (jpountz) [inline lucene/core/src/test/org/apache/lucene/index/TestFieldInfos.java]: We usually don't use this annotation and only rename on naming conventions: all methods that start with `test` are tested.

File: lucene/core/src/test/org/apache/lucene/index/TestFieldInfos.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  public void testFieldInfosSparse() throws Exception{
```
Comment (jpountz) [inline lucene/core/src/test/org/apache/lucene/index/TestFieldInfos.java]: Thanks for adding a test! Let's also add a similar test for empty and dense?

**Comment by mayya-sharipova** [comment]
@jpountz Adrien, thanks for your feedback. I tried to address it in the 2nd commit.

File: lucene/core/src/java/org/apache/lucene/index/FieldInfos.java [hunk 6]
```diff
@@ -63,21 +62,30 @@ public FieldInfos(FieldInfo[] infos) {
     boolean hasNorms = false;
     boolean hasDocValues = false;
     boolean hasPointValues = false;
-    
-    TreeMap<Integer, FieldInfo> byNumber = new TreeMap<>();
+
+    int size = 0; // number of elements in byNumberTemp
+    int capacity = 10; // byNumberTemp's capacity
+    FieldInfo[] byNumberTemp = new FieldInfo[capacity];
```
Comment (diegoceccarelli) [inline lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: what not using `ArrayList` ?

Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: as far as I'm concerned, I'm fine with either option: array or ArrayList.

File: lucene/core/src/java/org/apache/lucene/index/FieldInfos.java [hunk 7]
```diff
@@ -63,21 +62,30 @@ public FieldInfos(FieldInfo[] infos) {
     boolean hasNorms = false;
     boolean hasDocValues = false;
     boolean hasPointValues = false;
-    
-    TreeMap<Integer, FieldInfo> byNumber = new TreeMap<>();
+
+    int size = 0; // number of elements in byNumberTemp
+    int capacity = 10; // byNumberTemp's capacity
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: I think we can simplify a bit by removing these two local variables:
 - `capacity` can be replaced with byNumberTemp.length
 - `size` seems only useful to terminate early when creating a list of the field infos from byNumberTemp. But since ArrayUtil.grow is used, arrays are never more than ~1/8th larger than the number of slots that are actually used, so it doesn't buy much?

File: lucene/core/src/test/org/apache/lucene/index/TestFieldInfos.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    assertNull(fis2.fieldInfo(1));
+    assertEquals("f15", fis2.fieldInfo(15).name);
+    assertEquals("f16", fis2.fieldInfo(16).name);
+
+    // testing empty FieldInfos
+    assertNull(fis3.fieldInfo(0));
+    assertEquals(0, fis3.size());
+    Iterator<FieldInfo> it3 = fis3.iterator();
+    assertFalse(it3.hasNext());
+    dir.close();
```
Comment (jpountz) [inline lucene/core/src/test/org/apache/lucene/index/TestFieldInfos.java]: in each of these 3 cases, can you also do some lookups by name and by number?

File: lucene/core/src/java/org/apache/lucene/index/FieldInfos.java (hunk 6 above)
Comment (mayya-sharipova) [reply lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: @diegoceccarelli thanks for a suggestion. I considered `ArrayList` as well, but the code was even more complex. `byNumberTemp` could potentially be a sparse array, where some positions don't have values.  You also need to `ensureCapacity` of the ArrayList, and `add` null elements.

File: lucene/core/src/java/org/apache/lucene/index/FieldInfos.java (hunk 7 above)
Comment (mayya-sharipova) [reply lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: @jpountz Thank you for the suggestion on `capacity` - very nice one.  For the `size`, later we use this variable to create the final array: `byNumber = new FieldInfo[size]`. We can also do: `byNumber = new FieldInfo[byNumberTemp.length]`, but then `byNumber` may have some empty slots.

File: lucene/core/src/java/org/apache/lucene/index/FieldInfos.java (hunk 6 above)
Comment (diegoceccarelli) [reply lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: That's true :)  what about having an `ArrayList` and then a separate private method `ensureSize` ?

```
private static void ensureSize(ArrayList<?> list, int size) {
    list.ensureCapacity(size);
    while (list.size() < size) {
        list.add(null);
    }
}

```

In this way you will not have to care about the logic behind the resizing and you will have less variables around.

Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/index/FieldInfos.java]: This looks more complex to me than manipulating the array directly?
